ssl/quic/quic_fc.c

```diff
@@ -0,0 +1,329 @@
+    rxfc->is_fin            = 0;
+    return 1;
+}
+
+static void rxfc_start_epoch(QUIC_RXFC *rxfc)
+{
+    rxfc->epoch_start   = rxfc->now(rxfc->now_arg);
+    rxfc->esrwm         = rxfc->rwm;
+}
+
+static int on_rx_controlled_bytes(QUIC_RXFC *rxfc, uint64_t num_bytes)
+{
+    int ok = 1;
+    uint64_t credit = rxfc->cwm - rxfc->swm;
+
+    if (num_bytes > credit) {
+        ok = 0;
+        num_bytes = credit;
+        rxfc->error_code = QUIC_ERR_FLOW_CONTROL_ERROR;
+    }
+
+    rxfc->swm += num_bytes;
+    return ok;
+}
+
+int ossl_quic_rxfc_on_rx_stream_frame(QUIC_RXFC *rxfc, QUIC_RXFC *conn_rxfc,
+                                      uint64_t end, int is_fin)
+{
+    uint64_t delta;
+
+    if (!rxfc->is_stream || conn_rxfc == NULL || conn_rxfc->is_stream)
+        return 0;
+
+    if (rxfc->is_fin && ((is_fin && rxfc->hwm != end) || end > rxfc->hwm)) {
+        /* Stream size cannot change after the stream is finished */
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    if (is_fin)
+        rxfc->is_fin = 1;
+
+    if (end > rxfc->hwm) {
+        delta = end - rxfc->hwm;
+        rxfc->hwm = end;
+
+        on_rx_controlled_bytes(rxfc, delta);        /* result ignored */
+        on_rx_controlled_bytes(conn_rxfc, delta);   /* result ignored */
+    } else if (end < rxfc->hwm && is_fin) {
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    return 1;
+}
+
+/* threshold = 3/4 */
+#define WINDOW_THRESHOLD_NUM 3
+#define WINDOW_THRESHOLD_DEN 4
+
+static int rxfc_cwm_bump_desired(QUIC_RXFC *rxfc)
+{
+    uint64_t window_rem = rxfc->cwm - rxfc->rwm;
+    uint64_t threshold
+        = (rxfc->cur_window_size * WINDOW_THRESHOLD_NUM) / WINDOW_THRESHOLD_DEN;
+
+    return window_rem <= threshold;
+}
+
+static int rxfc_should_bump_window_size(QUIC_RXFC *rxfc, OSSL_TIME rtt)
+{
+    /*
+     * dt: time since start of epoch
+     * b:  bytes of window consumed since start of epoch
+     * dw: proportion of window consumed since start of epoch
+     * T_window: time it will take to use up the entire window, based on dt, dw
+     * RTT: The current estimated RTT.
+     *
+     * b        = rwm - esrwm
+     * dw       = b / window_size
+     * T_window = dt / dw
+     * T_window = dt / (b / window_size)
+     * T_window = (dt * window_size) / b
+     *
+     * We bump the window size if T_window < 4 * RTT.
+     *
+     * We leave the division by b on the LHS to reduce the risk of overflowing
+     * our 64-bit nanosecond representation, which will afford plenty of
+     * precision left over after the division anyway.
+     */
+    uint64_t  b = rxfc->rwm - rxfc->esrwm;
+    OSSL_TIME now, dt, t_window;
+
+    if (b == 0)
+        return 0;
+
+    now      = rxfc->now(rxfc->now_arg);
+    dt       = ossl_time_subtract(now, rxfc->epoch_start);
+    t_window = ossl_time_divide(ossl_time_multiply(dt, rxfc->cur_window_size), b);
```

Comment:
I wasn't expecting to need a fused */ operation for times.
I'll include it as part of the OSSL_TIME work I'm doing.

Reply:
`ossl_time_muldiv()` exists now, best use it here.